fhir-persistence-cos/src/main/java/com/ibm/fhir/persistence/cos/client/COSPayloadClient.java

```diff
@@ -0,0 +1,232 @@
+import com.ibm.cloud.objectstorage.auth.AWSCredentials;
+import com.ibm.cloud.objectstorage.auth.AWSStaticCredentialsProvider;
+import com.ibm.cloud.objectstorage.auth.BasicAWSCredentials;
+import com.ibm.cloud.objectstorage.client.builder.AwsClientBuilder.EndpointConfiguration;
+import com.ibm.cloud.objectstorage.oauth.BasicIBMOAuthCredentials;
+import com.ibm.cloud.objectstorage.services.s3.AmazonS3;
+import com.ibm.cloud.objectstorage.services.s3.AmazonS3ClientBuilder;
+import com.ibm.cloud.objectstorage.services.s3.model.CreateBucketRequest;
+import com.ibm.cloud.objectstorage.services.s3.model.DeleteObjectRequest;
+import com.ibm.cloud.objectstorage.services.s3.model.GetObjectRequest;
+import com.ibm.cloud.objectstorage.services.s3.model.ObjectMetadata;
+import com.ibm.cloud.objectstorage.services.s3.model.PutObjectRequest;
+import com.ibm.cloud.objectstorage.services.s3.model.PutObjectResult;
+import com.ibm.cloud.objectstorage.services.s3.model.S3Object;
+import com.ibm.cloud.objectstorage.services.s3.model.S3ObjectInputStream;
+import com.ibm.fhir.persistence.exception.FHIRPersistenceException;
+import com.ibm.fhir.persistence.jdbc.exception.FHIRPersistenceDataAccessException;
+import com.ibm.fhir.persistence.util.InputOutputByteStream;
+
+/**
+ * Encapsulates the AmazonS3 client for handling COS payloads
+ */
+public class COSPayloadClient {
+    private static final Logger logger = Logger.getLogger(COSPayloadClient.class.getName());
+
+    // The client connection established by calling #connect()
+    private final AmazonS3 client;
+
+    private final CosPropertyGroupAdapter propertyAdapter;
+
+    // The tenant this client is tied to
+    private final String tenantId;
+
+    static {
+        SDKGlobalConfiguration.IAM_ENDPOINT = "https://iam.cloud.ibm.com/oidc/token";
+    }
+
+    /**
+     * Public constructor
+     * @param cosProperties
+     */
+    public COSPayloadClient(String tenantId, CosPropertyGroupAdapter propertyAdapter) {
+
+        this.tenantId = tenantId;
+        this.propertyAdapter = propertyAdapter;
+
+        AWSCredentials credentials;
+        if (propertyAdapter.isCredentialIBM()) {
+            credentials = new BasicIBMOAuthCredentials(propertyAdapter.getApiKey(), propertyAdapter.getSrvInstId());
+        } else {
+            credentials = new BasicAWSCredentials(propertyAdapter.getApiKey(), propertyAdapter.getSrvInstId());
+        }
+
+        ClientConfiguration clientConfig = new ClientConfiguration()
+                .withRequestTimeout(propertyAdapter.getRequestTimeout())
+                .withTcpKeepAlive(true)
+                .withSocketTimeout(propertyAdapter.getSocketTimeout());
+
+        // connect to the configured endpoint/location
+        this.client = AmazonS3ClientBuilder.standard().withCredentials(new AWSStaticCredentialsProvider(credentials))
+            .withEndpointConfiguration(new EndpointConfiguration(propertyAdapter.getEndpointUrl(), propertyAdapter.getLocation()))
```

Comment:
Endpoint can be a virtual host, and it's not accounted for.  It's worth just documenting the fact that VirtualHosts are not supported.